### PR TITLE
Make apps handle the order logic

### PR DESCRIPTION
**Pull request by skjnldsv**
Let the app return a lower order if the url matches its requirements!

Will be cleaner than checking against a weird id!

- [x] Files
- [x] Comments
- [x] Settings
- [x] Contacts
- [x] Events
- [x] Tasks
- [ ] Talk https://github.com/nextcloud/spreed/pull/3992

### CODE REVIEW

File: core/Controller/UnifiedSearchController.php [hunk 1]
```diff
@@ -54,10 +54,14 @@
 	/**
 	 * @NoAdminRequired
 	 * @NoCSRFRequired
+	 * 
+	 * @param string $from the url the user is currently at
```
Comment (nickvergessen) [inline core/Controller/UnifiedSearchController.php:58]: `string $from`

Comment (skjnldsv) [reply core/Controller/UnifiedSearchController.php:58]: yes?

File: core/Controller/UnifiedSearchController.php [hunk 2]
```diff
@@ -54,10 +54,14 @@ public function __construct(IRequest $request,
...
+	 * 
+	 * @return JSONResponse
 	 */
-	public function getProviders(): JSONResponse {
+	public function getProviders($from): JSONResponse {
```
Comment (nickvergessen) [inline core/Controller/UnifiedSearchController.php:62]: My comment should go here... The mobile app messes it up

Comment (skjnldsv) [reply core/Controller/UnifiedSearchController.php:62]: :+1:

File: apps/comments/lib/Search/CommentsSearchProvider.php [hunk 3]
```diff
@@ -77,7 +77,7 @@ public function getName(): string {
 	/**
 	 * @inheritDoc
 	 */
-	public function getOrder(): int {
+	public function getOrder($from): int {
```
Comment (nickvergessen) [inline apps/comments/lib/Search/CommentsSearchProvider.php:80]:
```suggestion
	public function getOrder(string $from): int {
```

same on all other implementations

File: apps/settings/lib/Search/SectionSearch.php [hunk 4]
```diff
@@ -75,7 +75,11 @@ public function getName(): string {
 	/**
 	 * @inheritDoc
 	 */
-	public function getOrder(): int {
+	public function getOrder(string $from): int {
+		if (strpos($from, $this->urlGenerator->linkToRoute('settings.PersonalSettings.index') === 0)
+		|| strpos($from, $this->urlGenerator->linkToRoute('settings.AdminSettings.index')) === 0) {
+			return -1;
```
Comment (skjnldsv) [inline apps/settings/lib/Search/SectionSearch.php:81]: @nickvergessen what are your thoughts on this? :thinking:

Comment (skjnldsv) [reply apps/settings/lib/Search/SectionSearch.php:81]: Can lead to issues with `/settings/user` and `/settings/users` for example :thinking:

Comment (skjnldsv) [reply apps/settings/lib/Search/SectionSearch.php:81]: Ideally we should use https://github.com/nextcloud/server/blob/caff1023ea72bb2ea94130e18a2a6e2ccf819e5f/lib/public/Route/IRouter.php#L106
But it's deprecated :thinking:

File: core/Controller/UnifiedSearchController.php [hunk 5]
```diff
@@ -129,7 +143,7 @@ protected function getRouteInformation(string $url): array {
 		}
 
 		return [
-			$route,
+			$routeStr,
```
Comment (nickvergessen) [inline core/Controller/UnifiedSearchController.php:146]: potentially undefined, also needs adjustment in:
```php
	protected function getRouteInformation(string $url): array {
		$route = '';
```

**Comment by skjnldsv** [comment]
Done, let's wait for #22099 to me merged

**Comment by skjnldsv** [comment]
Ready to review, compile bot incoming! :rocket:

**Review by MorrisJobke — APPROVED** [review]
Code looks good 👍

**Comment by skjnldsv** [comment]
All green :green_heart:

**Comment by MorrisJobke** [comment]
> All green 💚

Except a second approval 😝

**Comment by skjnldsv** [comment]
> Except for the second approval

I'm trying to lure some people here :sob:

**Review by nickvergessen — APPROVED** [review]

**Comment by skjnldsv** [comment]
:hugs:

**Pull request merged**